gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +307,306 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
-  inline Key key5() const { return keys_[4]; }
-  inline Key key6() const { return keys_[5]; }
-
-  /** Calls the 6-key specific version of evaluateError, which is pure virtual
-   * so must be implemented in the derived class. */
-  Vector unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H = boost::none) const override {
-    if(this->active(x)) {
-      if(H)
-        return evaluateError(x.at<X1>(keys_[0]), x.at<X2>(keys_[1]), x.at<X3>(keys_[2]), x.at<X4>(keys_[3]), x.at<X5>(keys_[4]), x.at<X6>(keys_[5]), (*H)[0], (*H)[1], (*H)[2], (*H)[3], (*H)[4], (*H)[5]);
-      else
-        return evaluateError(x.at<X1>(keys_[0]), x.at<X2>(keys_[1]), x.at<X3>(keys_[2]), x.at<X4>(keys_[3]), x.at<X5>(keys_[4]), x.at<X6>(keys_[5]));
+ private:
+  /** Pack expansion with index_sequence template pattern, used to index into
+   * `keys_` and `H`
+   */
+  template <std::size_t... Inds>
+  inline Vector unwhitenedError(
+      boost::mp11::index_sequence<Inds...>,  //
+      const Values& x,
+      boost::optional<std::vector<Matrix>&> H = boost::none) const {
+    if (this->active(x)) {
+      if (H) {
+        return evaluateError(x.at<VALUES>(keys_[Inds])..., (*H)[Inds]...);
+      } else {
+        return evaluateError(x.at<VALUES>(keys_[Inds])...);
+      }
     } else {
       return Vector::Zero(this->dim());
     }
   }
 
-  /**
-   *  Override this method to finish implementing a 6-way factor.
-   *  If any of the optional Matrix reference arguments are specified, it should compute
-   *  both the function evaluation and its derivative(s) in X1 (and/or X2, X3).
-   */
-  virtual Vector
-  evaluateError(const X1&, const X2&, const X3&, const X4&, const X5&, const X6&,
-      boost::optional<Matrix&> H1 = boost::none,
-      boost::optional<Matrix&> H2 = boost::none,
-      boost::optional<Matrix&> H3 = boost::none,
-      boost::optional<Matrix&> H4 = boost::none,
-      boost::optional<Matrix&> H5 = boost::none,
-      boost::optional<Matrix&> H6 = boost::none) const = 0;
-
-private:
-
   /** Serialization function */
   friend class boost::serialization::access;
-  template<class ARCHIVE>
-  void serialize(ARCHIVE & ar, const unsigned int /*version*/) {
-    ar & boost::serialization::make_nvp("NoiseModelFactor",
-        boost::serialization::base_object<Base>(*this));
+  template <class ARCHIVE>
+  void serialize(ARCHIVE& ar, const unsigned int /*version*/) {
+    ar& boost::serialization::make_nvp(
+        "NoiseModelFactor", boost::serialization::base_object<Base>(*this));
   }
-}; // \class NoiseModelFactor6
+};  // \class NoiseModelFactorN
 
 /* ************************************************************************* */
+/** @deprecated: use NoiseModelFactorN
+ * Convenient base classes for creating your own NoiseModelFactors with 1-6
+ * variables.  To derive from these classes, implement evaluateError().
+ *
+ * Note: we cannot use `using NoiseModelFactor1 = NoiseModelFactorN<VALUE>` due
```

Comment:
Can you comment on that ?

Reply:
If you mean can I explain more in detail here, yes:

In c++, a derived class can refer to its base class either _with_ or _without_ template parameters.  This is a syntactical convenience: you can always refer to yourself (or your visible ancestors) from within your class scope.  See [injected-class-name](https://en.cppreference.com/w/cpp/language/injected-class-name) for full description.

Here's a simple example, where `Derived` can either use the `B` form _or_ the `B<T>` form.

```c++
template <typename T>
struct Base {
  using Type = T;
};

template <typename T>
struct Derived : public Base<T>{
  using Type1 = Base::Type;
  using Type2 = Base::<T>::Type;
};
```

As a note, this means instead of the common pattern
```c++
using Base = NoiseModelFactor4<T1, T2, T3, T4>;
```
I'm pretty sure we could instead just do
```c++
using Base = NoiseModelFactor4;
```

The reason this is problematic is because if you do
```c++
template <typename T>
using Derived = Base<T>
```

Just a weird quirk that anyone who inherits from `Derived` does NOT get class name injection so they _must_ use `Base<T>` and _cannot_ use just `Base`.  I'm not sure why the spec was written this way.

In one test file in the repo, someone does use just `NoiseModelFactor6` in a derived class without specifying the template parameters.  We could just change the unit test, but then that would be in conflict with:
> has to be 100% compatible of course.

Also, if we try doing something tricky like

```c++
class NoiseModelFactorN<...> {
  ...
  using NoiseModelFactor4 = This;
};
```

Then the `NoiseModelFactor4<T>` syntax doesn't work in the derived class.  So basically I'm almost certain there isn't a way to be able to use both syntaxes: `NoiseModelFactor4` and `NoiseModelFactor4<T1, T2, T3, T4>` using only `using`.

An alternative is to define structs:

```c++
template <...>
struct NoiseModelFactor4 : public NoiseModelFactorN<...> {
 public:
  using NoiseModelFactorN<...>::NoiseModelFactorN;
  using This = NoiseModelFactor4<...>;
};
```

But I don't see this as being significantly better than the typedefs.

One advantage of structs is that we could then move the deprecated `keyX()` functions and `using X` into there.